### ISSUE TITLE
Indexing relations in search

### ISSUE THREAD

**Issue by kaedroho** [torchbox/wagtail#459]
This is useful for filtering/searching on related/child objects.

I suggest adding a new search field type called ```RelatedFields```.

**Examples**

Indexing an authors name/date_of_birth fields

```python
class Book(models.Model, indexed.Indexed):
    ...

    search_fields = [
        indexed.SearchField('title'),
        indexed.FilterField('published_date'),

        indexed.RelatedFields('author', [
            indexed.SearchField('name'),
            indexed.FilterField('date_of_birth'),
        ]),
    ]

# Search books where their author was born after 1950
# Both the book title and the authors name will be searched
>>> Book.objects.filter(author__date_of_birth__gt=date(1950, 1, 1)).search("Hello")
```

The ```RelatedFields``` would work on reverse relationships as well:

```python
class Author(models.Model, indexed.Indexed):
    ...

    search_fields = [
        indexed.SearchField('name'),
        indexed.FilterField('date_of_birth'),

        indexed.RelatedFields('published_books', [
            indexed.SearchField('title'),
            indexed.FilterField('published_date'),
        ]),
    ]

# Search authors who have published a book after 1950
# Both the authors name and their book titles will be searched
>>> Author.objects.filter(book__published_date__gt=date(1950, 1, 1)).search("Hello")
```

It's also possible to do this (this would ignore the related fields defined on the book model):

```python
class Author(models.Model, indexed.Indexed):
    ...

    search_fields = [
        indexed.SearchField('name'),
        indexed.FilterField('date_of_birth'),

        indexed.RelatedFields('published_books', Book.search_fields),
    ]
```

**Comment by kaedroho** [comment]
Had a look into this yesterday. The indexing part looks quite simple. In Elasticsearch, we can just stick the related objects into nested documents.

Filtering on related fields while performing a search query looks like it might be tricky due to the way Django represents these in a QuerySet (they're converted to a Join object https://github.com/django/django/blob/master/django/db/models/sql/datastructures.py#L28-L118).

I think we can implement just indexing first and figure out a way around the filtering issue at a later date.

**Comment by SalahAdDin** [comment]
This is brutal, thank you very much, awesome.

**Comment by kaedroho** [comment]
Just implemented the syntax/indexing code: https://github.com/torchbox/wagtail/pull/1491

**Comment by gasman** [comment]
Implemented in 2c765a7462a8270cb471f74d632fc1c2a1aeb58a + parents
